### ISSUE TITLE
Anybody can subscribe other people's emails

### ISSUE THREAD

**Issue by Yajo** [odoo/odoo#33234]
Impacted versions: 8-12

Steps to reproduce:

1. Install any module that lets you subscribe to something in odoo publicly. Examples:

   - website_mass_mailing
   - website_blog
   - website_slides

   There are possibly others, but these come to my mind now.

1. Make sure a subscription widget exists. I.e. for `website_mass_mailing` you have to drag it to a web page.

1. Go to the website in a private window and subscribe anybody's email.

Current behavior:

You are allowed to do that as many times as you want. You are able to subscribe anybody anywhere. **Exploiting this with a simple js script to subscribe a ton of emails would be a piece of cake.**

Expected behavior:

A validation email should be sent to the subscriber.

Additional info:

:thinking: IANAL, but I think [GDPR art. 4.1](https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:32016R0679#d1e1489-1-1) says an email could be a personal data:

> 'personal data’ means any information  relating to an identified or identifiable natural person (‘data  subject’); an identifiable natural person is one who can be identified,  directly or indirectly, in particular by reference to an identifier such  as a name, an identification number, location data, **an online  identifier** or to one or more factors specific to the physical,  physiological, genetic, mental, economic, cultural or social identity of  that natural person;

...  and I also think that sending an email to somebody who didn't subscribe is spam (based on common sense; after all, email verification has been in mailing lists like forever, much before the GDPR came along) and is not any of the lawfullness processings stated in [GDPR art. 6](https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:32016R0679#d1e1888-1-1) (based on my limited understanding of legal texts).

So maybe this is a legal problem, maybe it's a security problem (the security team told me 1 year ago it isn't), maybe a privacy problem, maybe a feature request , or a bug... Or maybe I'm missing some important points... Not sure, but anyways AFAICS it should be fixed.

@Tecnativa @rafaelbn

**Comment by pedrobaeza** [comment]
@odony what do you think?

**Comment by Yenthe666** [comment]
I get what you guys mean but to put it simply: how would you intent to prevent this or be sure that somebody is not spam adding? You could add a limit on the amount of requests/registers maybe but thats about it, no?

**Comment by Yajo** [comment]
A good workflow would be:

1. You subscribe someone (usually you).
1. The subscription is in `state=unconfirmed`.
1. A verification email is sent: "You have been subscribed to X. Was it you? Confirm it clicking here: BUTTON. Otherwise, just ignore this email"
1. When confirmed, `state=confirmed` and starts getting emails.

The limits is another good option, although complementary IMHO.

**Comment by pedrobaeza** [comment]
Well, one reasonable measure is to have a pool of unconfirmed emails, for at least not flooding/spamming an email without that confirmation when being subscribed to several lists.

**Comment by Yenthe666** [comment]
Ahh I misunderstood your initial issue. I assumed you wanted to stop people from even entering/sending it. An extra verification check with an e-mail sounds like a good addition to me, along with being able to subscribe and unsubscribe on different mailing lists too.

**Comment by Yajo** [comment]
@pedrobaeza Yes, although `website_mass_mailing` is a specific case for mass mailings. Other models like blog posts or website slides just inherit from `mail.thread` and leverage its features. Maybe a good solution there would be to create partners with `opt_out=True` until confirmed from that email, so even if subscribed, they wouldn't get mails until confirmed. :thinking:

**Comment by Yenthe666** [comment]
@tde-banana-odoo might like to join the opt-in/opt-out discussion too probably.

**Comment by tde-banana-odoo** [comment]
We already has this discussion last year. It is indeed a good practice according to me. However it is not mandatory and Odoo choose not to develop it. What we did is that you are not subscribed by default to any newsletter / channel available in Odoo; you have to manually choose each spam source to follow. You can also easily blacklist your email, meaning you don't receive any mass email anymore whatever the source.

Subscribing someone else email to any web service you find is effectively doable. Lot of websites allow to subscribe whatever email. Indeed it may be annoying if someone puts your email in all newsletter-spamming websites he encounters. Currently Odoo choosed not to fight againt the giant spam machine.

Anyway we current models in master I think all pieces are present to develop a double opt-in module in a quite straight-forward way. I think you already have that find of module as we discussed about it last year. If there are pieces of code to improve in standard addons to ease integration feel free to propose a PR :) .

Cheers,

**Comment by sswapnesh** [comment]
> A verification email is sent: "You have been subscribed to X. Was it you? Confirm it clicking here: BUTTON. Otherwise, just ignore this email"

> When confirmed, state=confirmed and starts getting emails.

if I am not mistaken [Odoo](odoo.com) is using same workflow.

**Comment by Yajo** [comment]
I currently don't have any code to share. It would be best to have the design first. The closest thing we have developed is the [`privacy_consent`](https://github.com/OCA/data-protection/tree/10.0/privacy_consent) module, which asks for acceptance and records user actions.



> Anyway we current models in master I think all pieces are present to develop a double opt-in module in a quite straight-forward way.

I'm not aware of such pieces, could you specify? :thinking:

**Comment by jans23** [comment]
The issue I see is that one attacker could send subscription requests (not successfully subscribe) to as many emails as she likes. The consequences we were experiencing are a very slow system during such attacks and potentially receiving spam filters would rate our address negatively.

To avoid such attacks a throttling logic would be good to prevent one IP address to subscribe an indefinite amount of addresses or use a captcha. Is any such solution available?

I see that this point is a bit different from the original issue description but it's still very much related. Please let me know if you prefer me filling a separate issue ticket.

**Comment by Yajo** [comment]
Then maybe it means that in more recent versions, Odoo finally implemented the flow explained in https://github.com/odoo/odoo/issues/33234#issuecomment-490400511?

**Comment by jans23** [comment]
I don't remember the stock behaviour but for Odoo 12.0 we implemented [this module for double confirmation](https://github.com/Nitrokey/odoo-modules/tree/12.0/mass_mailing_double_opt_in).
